docs(css): FF127 - `calc()` support for color channels

dipikabh opened the pull request:
<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

In FF127, the CSS [`calc()`](https://developer.mozilla.org/en-US/docs/Web/CSS/calc) function can parse color channels.

The support is currently available behind the preference `layout.css.relative-color-syntax.enabled`, which is enabled by default in Nightly.

**Related Firefox bug**: [[css-color] calc() should also parse color component symbols (r, g, b, h, s, l, etc.)](https://bugzilla.mozilla.org/show_bug.cgi?id=1889561)

### Motivation

This PR adds the example "Adjusting colors dynamically" to demo the parsing of color channels by `calc()`.

### Additional details

Other updates done in this PR:
- Updated the "Syntax" section to align with the [CSS function page template](https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Page_structures/Page_types/CSS_function_page_template#syntax). Also added a syntax example to show parsing of color channels by `calc()`.
- Updated the H3 "[Notes](https://developer.mozilla.org/en-US/docs/Web/CSS/calc#notes)" section under "Syntax" to H2 "Description".
- Moved the "[Usage with integers](https://developer.mozilla.org/en-US/docs/Web/CSS/calc#usage_with_integers)" section as a bullet point under "Description".
- Updated the "[Formal syntax](https://developer.mozilla.org/en-US/docs/Web/CSS/calc#formal_syntax)" section from H3 to H2 as per the template.

### Related issues and pull requests

Doc issue tracker: https://github.com/mdn/content/issues/33586
BCD update: https://github.com/mdn/browser-compat-data/pull/23324
Firefox release note update: https://github.com/mdn/content/pull/34042

dipikabh committed:
updates calc() for color support

bsmth commented on files/en-us/web/css/calc/index.md:
```diff
@@ -14,8 +14,14 @@ The **`calc()`** [CSS](/en-US/docs/Web/CSS) [function](/en-US/docs/Web/CSS/CSS_F
-/* property: calc(expression) */
-width: calc(100% - 80px);
+/* calc(expression) */
+calc(100% - 80px)
```
Should we add this:


```suggestion
calc(100% - 80px)

/* Expression with a CSS function */
calc(100px * sin(pi / 2))

/* Expression containing a variable */
calc(var(--hue) + 180)
```

dipikabh replied:
This is great, thanks! Adding

bsmth commented on files/en-us/web/css/calc/index.md:
```diff
@@ -14,8 +14,14 @@ The **`calc()`** [CSS](/en-US/docs/Web/CSS) [function](/en-US/docs/Web/CSS/CSS_F
+/* calc(expression) */
+calc(100% - 80px)
+
+/* Expression with color channel */
+calc(l * 0.8)
+calc(h + 180)
+calc(b / 2)
+calc(r * .3 + g * .59 + b * .11)
```
What do you think about this?

```suggestion
/* Expression with color channels */
calc(alpha * 0.8)
calc(r / 2) calc(g + 25) calc(b + 175)
```

For the last one, I think it needs to be in this format:

```css
  background-color: rgb(from blue calc(r * .3) calc(g * .59) calc(b * .11));
```

bsmth replied:
And we're talking about use in relative colors, so this is more explicit, but it's verbose

```
/* Expression with color channels in relative colors */
```

dipikabh replied:
- The "Syntax" guidelines for CSS function pages were updated: https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Page_structures/Page_types/CSS_function_page_template#syntax, and so now we don't mention the property in which the function is used or the ending semi-colon. I realize there are loads of other CSS function pages where we still need to implement this new guideline.

- Agree with your suggestion about explicitly stating use in relative colors

bsmth commented on files/en-us/web/css/calc/index.md:
```diff
@@ -45,8 +51,9 @@ In addition, the following notes apply:
 - For lengths, you can't use `0` to mean `0px` (or another length unit); instead, you must use the version with the unit: `margin-top: calc(0px + 20px);` is valid, while `margin-top: calc(0 + 20px);` is invalid.
 - The `calc()` function cannot directly substitute the numeric value for percentage types; for instance `calc(100 / 4)%` is invalid, while `calc(100% / 4)` is valid.
 - Current implementations require that for the `*` and `/` operators, one of the operands has to be unitless. For `/`, the right operand must be unitless. For example `font-size: calc(1.25rem / 1.25)` is valid but `font-size: calc(1.25rem / 125%)` is invalid.
+- When `calc()` is used where an {{cssxref("&lt;integer&gt;")}} is expected, the value will be rounded to the nearest integer. For example, `z-index: calc(3 / 2);` will result in a `z-index` value of `2`.
```
I guess it's rounded up, is it?

bsmth replied:
Well, I guess rounded to the nearest integer is fine, but if halfway, it's rounded up:

* `z-index: calc(1.4);` = 1
* `z-index: calc(1.5);` = 2
* `z-index: calc(1.6);` = 2

dipikabh replied:
Good point. I also found this reference in the spec: [calc type checking](https://drafts.csswg.org/css-values/#calc-type-checking) leading to [round to the nearest integer](https://drafts.csswg.org/css-values/#css-round-to-the-nearest-integer).
I've updated the text to (removed z-index from the example as it was distracting):

> - When `calc()` is used where an {{cssxref("&lt;integer&gt;")}} is expected, the value will be rounded to the nearest integer. So, `calc(1.4)` will result in a value of `1`. If the fractional part of the value is exactly `0.5`, the value will be rounded up. For example, `calc(1.5)` will result in a value of `2`, while `calc(-1.5)` will round to `-1`.

bsmth commented on files/en-us/web/css/calc/index.md:
````diff
@@ -149,7 +144,58 @@ You can also use `calc()` with [CSS variables](/en-US/docs/Web/CSS/CSS_cascading
+  This text color is adjusted to decrease the red channel by 200.
+</p>
+<p class="double">This text color is adjusted to double the green channel.</p>
+<p class="half">
+  This text color is dynamically adjusted to halve the alpha channel.
+</p>
+```
+
+```css
+:root {
+  --r: 255;
+  --g: 100;
+  --b: 100;
+  --a: 1;
+}
+
+.original {
+  color: rgb(var(--r) var(--g) var(--b));
+}
+
+.plus {
+  color: rgb(var(--r) var(--g) calc(var(--b) + 100));
+}
+
+.minus {
+  color: rgb(calc(var(--r) - 200) var(--g) var(--b));
+}
+
+.double {
+  color: rgb(var(--r) calc(var(--g) * 2) var(--b));
+}
+
+.half {
+  color: rgb(var(--r) var(--g) var(--b) / calc(var(--a) / 2));
+}
+```
+
+{{EmbedLiveSample('Adjusting_colors_dynamically', '700', '200')}}
+
+Refer to the [Using math functions](/en-US/docs/Web/CSS/CSS_colors/Relative_colors#using_math_functions) section to see how the `calc()` function is used with relative colors.
````
What do you think about this?

* un-hide the HTML
* Make the variables more explicit (`--red` instead of `--r`)
* Reduce repeated text in example
* Make the `<p>` text a little larger

````suggestion
### Adjusting colors with variables

The `calc()` function can be used with colors to adjust individual color channels.
In the example below, CSS variables are defined for the red, green, blue, and alpha channels of the [`rgb()`](/en-US/docs/Web/CSS/color_value/rgb) color function.
A base color is defined using these variables, and in each paragraph, `calc()` is used to expand or dynamically adjust the variables for each color channel.

```html
<p class="original">This text has the original color.</p>
<p class="plus">Increasing the blue channel by 100.</p>
<p class="minus">Decreasing the red channel by 200.</p>
<p class="double">Doubling the amount of green in the original color.</p>
<p class="half">Halving the alpha (transparency) channel.</p>
```

```css hidden
p {
  font-family: monospace;
  font-size: 16px;
}
```

```css
:root {
  --red: 255;
  --green: 100;
  --blue: 100;
  --alpha: 1;
}

.original {
  color: rgb(var(--red) var(--green) var(--blue));
}

.plus {
  color: rgb(var(--red) var(--green) calc(var(--blue) + 100));
}

.minus {
  color: rgb(calc(var(--red) - 200) var(--green) var(--blue));
}

.double {
  color: rgb(var(--red) calc(var(--green) * 2) var(--blue));
}

.half {
  color: rgb(var(--red) var(--green) var(--blue) / calc(var(--alpha) / 2));
}
```

{{EmbedLiveSample('adjusting_colors_with_variables', '700', '200')}}
````

bsmth replied:
I feel like we can also add a relative colors example seeing as we're adding the syntax here, and it's a good comparison to include. What do you think about this:

````md
### Adjusting color channels in relative colors

The `calc()` function can adjust individual color channels in relative colors without the need for storing color channel values as variables.
In the example below, the first paragraph uses a [named color](/en-US/docs/Web/CSS/named-color).
In the three paragraphs that follow, `calc()` is used with the [`rgb()`](/en-US/docs/Web/CSS/color_value/rgb) and [`hsl()`](/en-US/docs/Web/CSS/color_value/hsl) functions to adjust the values of each color component relative to the original named color.

```html
<p class="original">This text has the original color.</p>
<p class="plus-blue">Increasing the blue channel by 40.</p>
<p class="semi-transparent">Halving the alpha (transparency) channel.</p>
<p class="lower-saturation">Decreasing the saturation by 40.</p>
<p class="inverting-hue">Adding 180 degrees to the hue angle.</p>
```

```css hidden
p {
  font-family: monospace;
  font-size: 16px;
}
```

```css
.original {
  color: cornflowerblue;
}

.plus-blue {
  color: rgb(from cornflowerblue r g calc(b + 40));
}

.semi-transparent {
  color: rgb(from cornflowerblue r g b / calc(alpha / 2));
}

.lower-saturation {
  color: hsl(from cornflowerblue h calc(s - 40) l);
}

.inverting-hue {
  color: hsl(from cornflowerblue calc(h + 180) s l);
}
```

{{EmbedLiveSample('adjusting_color_channels_in_relative_colors', '700', '200')}}
````

dipikabh replied:
I'm going to delete the example I added. It is flawed because it did not really demo the color channels in `calc()`, which was what my intention. Sorry for the work you put in to improve it.

We definitely should add an example to show the use of `calc()` in relative colors, which is what you've suggested. I've taken your suggestion and modified it a bit (like used a different color and changed some values).

estelle commented:
I suggest linking the term "relative colors" to the relative colors guide, either next to the math functions link to that page or in the see also.

dipikabh committed:
fixes review comments

dipikabh commented:
Thanks a lot for the excellent review, @bsmth.

I was somewhat off the mark regarding `calc()`'s support for parsing color channels. This ability of `calc()` is relevant only in the context of relative colors.

Please see the following updates I've made to better reflect `calc()`s enhancement:
- In the [first para](https://pr34046.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/calc) on the page, I've added `<color-function>` to the list (it links to the color_value page):

> The **`calc()`** [CSS](/en-US/docs/Web/CSS) [function](/en-US/docs/Web/CSS/CSS_Functions) lets you perform calculations when specifying CSS property values. It can be used with {{cssxref("&lt;length&gt;")}}, {{cssxref("&lt;frequency&gt;")}}, {{cssxref("&lt;angle&gt;")}}, {{cssxref("&lt;time&gt;")}}, {{cssxref("&lt;percentage&gt;")}}, {{cssxref("&lt;number&gt;")}}, {{cssxref("&lt;integer&gt;")}}, and {{cssxref("color_value", "&lt;color-function&gt;")}} values.

- I've reorganized the info in the "[Description](https://pr34046.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/calc#description)" section to first present the general info about the function, then group the rules and best practices, and finally add a sub-section about `calc()`s support of color channels in relative colors ([Support for computing color channels in relative colors](https://pr34046.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/calc#support_for_computing_color_channels_in_relative_colors)).

For your reference, here's the link to the `calc()` spec that mentions relative colors: https://drafts.csswg.org/css-values/#calc-variables.

- I've replaced the example with what you suggested.

Would appreciate it if you can take a look again and let me know what you think.

dipikabh commented:
> I suggest linking the term "relative colors" to the relative colors guide

Thanks for the suggestion, @estelle 👍. The link is included in the [new subsection](https://pr34046.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/calc#support_for_computing_color_channels_in_relative_colors) in "Description".

estelle commented on files/en-us/web/css/calc/index.md:
```diff
@@ -7,15 +7,24 @@ browser-compat: css.types.calc
+calc(var(--hue) + 180)
+
+/* Expression with color channels in relative colors */
+rgb(from blue calc(r * .3) calc(g * .59) calc(b * .11))
```
blue is rgb(0 0 255 / 1 ) 
The calc here is  basically rgb((0*.3) (0 * .59) and (255 * .11)) or 
rgb(0 0 28).
But since your using an old color function with a limited gamut, the output here is `color(srgb 0 0 0.11)`. 

Suggest using a relative color using a function that doesn't return a `color()` function instead of the color function used. 
Also suggest using calc on one to two calcs, as most people will calc a hue or an alpha, but not change all the color channels. Even more useful doing a relative alpha or hue on a color custom property, but that might be beyond the scope of this page.

bsmth replied:
What about using lch()?

```suggestion
lch(from aquamarine l c calc(h + 180))
```

Could probably duplicate it below, also

estelle commented on files/en-us/web/css/calc/index.md:
```diff
@@ -149,7 +158,62 @@ You can also use `calc()` with [CSS variables](/en-US/docs/Web/CSS/CSS_cascading
+}
+
+.invert-hue {
+  color: hsl(from cornflowerblue calc(h + 180) s l);
```
Same issue. Querying the output color value via the [HTMLElement.style](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/style) property or the [CSSStyleDeclaration.getPropertyValue()](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleDeclaration/getPropertyValue) method returns the output color as a [color(srgb ...)](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/color) value. This will be really confusing if people use the devtools and hit control on the color swatch. The relative color tutorial covers this and has some good examples to inspire.

bsmth replied:
```suggestion
  color: lch(from aquamarine l c calc(h + 180));
```

bsmth reviewed (approved):
Thanks a lot! I'm going to leave a +1 pending the open suggestions / comments 👍🏻

bsmth committed:
Apply suggestions from code review

bsmth committed:
Merge branch 'main' into ff127-calc-color-content

dipikabh commented on files/en-us/web/css/calc/index.md:
````diff
@@ -149,7 +158,62 @@ You can also use `calc()` with [CSS variables](/en-US/docs/Web/CSS/CSS_cascading
+In the paragraphs that follow, `calc()` is used with the [`rgb()`](/en-US/docs/Web/CSS/color_value/rgb) and [`hsl()`](/en-US/docs/Web/CSS/color_value/hsl) functions to adjust the values of each color channel relative to the original named color.
+
+```html
+<p class="original">Original text color in <code>rebeccapurple</code></p>
+<p class="same-original">
+  Same as original color, but using <code>calc()</code>
+</p>
+<p class="increase-green">Green channel increased by 180</p>
+<p class="reduce-alpha">Alpha channel (transparency) reduced by half</p>
+<p class="reduce-saturation">Saturation reduced by 60</p>
+<p class="invert-hue">Hue angle increased by 180 degrees</p>
````
```suggestion
<p class="original">Original text color in rebeccapurple</p>
<p class="increase-hue">Hue increased by 80</p>
<p class="increase-lightness">Lightness increased by 20</p>
<p class="decrease-lightness">Lightness decreased by 10</p>
```

dipikabh commented on files/en-us/web/css/calc/index.md:
````diff
@@ -149,7 +158,62 @@ You can also use `calc()` with [CSS variables](/en-US/docs/Web/CSS/CSS_cascading
+```
+
+```css
+.original {
+  color: rebeccapurple;
+}
+
+.same-original {
+  color: rgb(from rebeccapurple calc(r) calc(g) calc(b));
+}
+
+.increase-green {
+  color: rgb(from rebeccapurple r calc(g + 180) b);
+}
+
+.reduce-alpha {
+  color: rgb(from rebeccapurple r g b / calc(alpha / 2));
+}
+
+.reduce-saturation {
+  color: hsl(from rebeccapurple h calc(s - 60) l);
+}
+
+.invert-hue {
+  color: lch(from aquamarine l c calc(h + 180));
+}
````
```suggestion
.original {
  color: rebeccapurple;
}

.increase-hue {
  color: lch(from rebeccapurple l c calc(h + 80));
}

.increase-lightness {
  color: lch(from rebeccapurple calc(l + 20) c h);
}

.decrease-lightness {
  color: lch(from rebeccapurple calc(l - 10) c h);
}
```

dipikabh reviewed (commented):
Addressing Estelle's feedback

dipikabh committed:
Apply suggestions from code review

dipikabh commented:
Thank you both for the feedback! I've addressed all the comments.
